core/src/main/java/org/mule/runtime/core/internal/profiling/AbstractProfilingService.java

```diff
@@ -0,0 +1,112 @@
+import org.mule.runtime.core.internal.profiling.notification.DefaultProfilingNotificationListener;
+
+import javax.inject.Inject;
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.mule.runtime.api.config.MuleRuntimeFeature.ENABLE_PROFILING_SERVICE;
+
+/**
+ * A {@link AbstractProfilingService} that discovers available {@link ProfilingDataConsumer}
+ *
+ * @since 4.4
+ */
+public abstract class AbstractProfilingService implements ProfilingService, Initialisable, Startable, Disposable {
+
+  @Inject
+  protected ServerNotificationManager notificationManager;
+
+  @Inject
+  private ServerNotificationHandler serverNotificationHandler;
+
+  FeatureFlaggingService featureFlags;
+
+  private final Set<NotificationListener<ProfilingNotification>> addedListeners = new HashSet<>();
+
+  @Override
+  public void initialise() throws InitialisationException {}
+
+  @Override
+  public void start() throws MuleException {
+    if (featureFlags.isEnabled(ENABLE_PROFILING_SERVICE)) {
+      registerNotificationListeners(getDiscoveryStrategy().discover());
+    }
+  }
+
+  private void registerNotificationListeners(Set<ProfilingDataConsumer<ProfilingEventContext>> profilerDataConsumers) {
+    profilerDataConsumers.forEach(this::registerNotificationListener);
+  }
+
+  private void registerNotificationListener(ProfilingDataConsumer<ProfilingEventContext> profilingDataConsumer) {
+    NotificationListener<ProfilingNotification> profilingNotificationListener =
+        new DefaultProfilingNotificationListener(profilingDataConsumer);
+    notificationManager.addListenerSubscription(profilingNotificationListener, pn -> filterByAction(profilingDataConsumer, pn));
+    addedListeners.add(profilingNotificationListener);
+  }
+
+  private boolean filterByAction(ProfilingDataConsumer<ProfilingEventContext> profilerDataConsumer,
+                                 ProfilingNotification profilerNotification) {
+    return profilerDataConsumer.profilerEventTypes().stream()
+        .anyMatch(
+                  eventType -> eventType.getProfilingEventName()
+                      .equals(profilerNotification.getActionName()))
+        &&
+        profilerDataConsumer.selector().test(((ProfilingEventContext) profilerNotification.getSource()));
+  }
+
+  @Override
+  public void dispose() {
+    if (!notificationManager.isDisposed()) {
+      addedListeners.forEach(listener -> notificationManager.removeListener(listener));
+    }
+  }
+
+  protected abstract ProfilerDataConsumerDiscoveryStrategy getDiscoveryStrategy();
+
+  public <T extends ProfilingEventContext> void notifyEvent(T profilerEventContext, ProfilingEventType<T> action) {
+    serverNotificationHandler.fireNotification(new ProfilingNotification(profilerEventContext, action));
+  }
+
+  @Inject
+  public void setFeatureFlags(FeatureFlaggingService featureFlags) {
+    this.featureFlags = featureFlags;
+  }
+
+  /**
+   * Configures {@link FeatureFlaggingService} for MULE-19588.
```

Comment:
MULE-19588 says `Implement the API specs for the profiler`. Wouldn't it be clearer to say `Configures {@link FeatureFlaggingService} for the profiles functionality`?